File: app/order-book-dex/src/program/pdas.ts
import { PublicKey } from "@solana/web3.js";
import { PROGRAM_ID } from "./utils/constants";
import { BN } from "@coral-xyz/anchor";

const ORDER_BOOK_CONFIG_SEED = "order-book-config";
const ORDER_POSITION_CONFIG_SEED = "order-position-config";
const BUY_SEED = "buy-market-pointer";
const SELL_SEED = "sell-market-pointer";
const MARKET_POINTER_SEED = "market-pointer";
const ORDER_POSITION_SEED = "order-position";
const VAULT_ACCOUNT_SEED = "vault-account";

export const getOrderBookConfigPDA = (
  tokenMintA: PublicKey,
  tokenMintB: PublicKey
): PublicKey => {
  return PublicKey.findProgramAddressSync(
    [
      tokenMintA.toBuffer(),
      tokenMintB.toBuffer(),
      Buffer.from(ORDER_BOOK_CONFIG_SEED)
    ],
    PROGRAM_ID
  )[0];
};

export const getBuyMarketPointerPDA = (
  orderBookConfig: PublicKey
): PublicKey => {
  return PublicKey.findProgramAddressSync(
    [
      Buffer.from(BUY_SEED),
      orderBookConfig.toBuffer(),
      Buffer.from(MARKET_POINTER_SEED)
    ],
    PROGRAM_ID
  )[0];
};

export const getSellMarketPointerPDA = (
  orderBookConfig: PublicKey
): PublicKey => {
  return PublicKey.findProgramAddressSync(
    [
      Buffer.from(SELL_SEED),
      orderBookConfig.toBuffer(),
      Buffer.from(MARKET_POINTER_SEED)
    ],
    PROGRAM_ID
  )[0];
};

export const getVaultAccountPDA = (
  orderBookConfig: PublicKey,
  tokenMint: PublicKey,
  signer: PublicKey
): PublicKey => {
  return PublicKey.findProgramAddressSync(
    [
      orderBookConfig.toBuffer(),
      tokenMint.toBuffer(),
      signer.toBuffer(),
      Buffer.from(VAULT_ACCOUNT_SEED)
    ],
    PROGRAM_ID
  )[0];
};

export const getOrderPositionConfigPDA = (
  signer: PublicKey,
  orderBookConfig: PublicKey
): PublicKey => {
  return PublicKey.findProgramAddressSync(
    [
      signer.toBuffer(),
      orderBookConfig.toBuffer(),
      Buffer.from(ORDER_POSITION_CONFIG_SEED)
    ],
    PROGRAM_ID
  )[0];
};

export const getOrderPositionPDA = (
  nonce: BN,
  signer: PublicKey
): PublicKey => {
  return PublicKey.findProgramAddressSync(
    [
      nonce.toArrayLike(Buffer, "le", 8),
      signer.toBuffer(),
      Buffer.from(ORDER_POSITION_SEED)
    ],
    PROGRAM_ID
  )[0];
};

File: backend/src/transactions/pdas.rs
use order_book_dex::{constants::*, ID as program_id};
use solana_sdk::pubkey::Pubkey;

pub fn get_order_book_config_pda(token_mint_a: Pubkey, token_mint_b: Pubkey) -> Pubkey {
    Pubkey::find_program_address(
        &[
            token_mint_a.as_ref(),
            token_mint_b.as_ref(),
            ORDER_BOOK_CONFIG_SEED.as_bytes(),
        ],
        &program_id,
    )
    .0
}

pub fn get_buy_market_pointer_pda(order_book_config: Pubkey) -> Pubkey {
    Pubkey::find_program_address(
        &[
            BUY_SEED.as_bytes(),
            order_book_config.as_ref(),
            MARKET_POINTER_SEED.as_bytes(),
        ],
        &program_id,
    )
    .0
}

pub fn get_sell_market_pointer_pda(order_book_config: Pubkey) -> Pubkey {
    Pubkey::find_program_address(
        &[
            SELL_SEED.as_bytes(),
            order_book_config.as_ref(),
            MARKET_POINTER_SEED.as_bytes(),
        ],
        &program_id,
    )
    .0
}

pub fn get_vault_account_pda(
    order_book_config: Pubkey,
    token_mint: Pubkey,
    signer: Pubkey,
) -> Pubkey {
    Pubkey::find_program_address(
        &[
            order_book_config.as_ref(),
            token_mint.as_ref(),
            signer.as_ref(),
            VAULT_ACCOUNT_SEED.as_bytes(),
        ],
        &program_id,
    )
    .0
}

pub fn get_order_position_config_pda(signer: Pubkey, order_book_config: Pubkey) -> Pubkey {
    Pubkey::find_program_address(
        &[
            signer.as_ref(),
            order_book_config.as_ref(),
            ORDER_POSITION_CONFIG_SEED.as_bytes(),
        ],
        &program_id,
    )
    .0
}

pub fn get_order_position_pda(nonce: u64, position_config: Pubkey, signer: Pubkey) -> Pubkey {
    let a = Pubkey::find_program_address(
        &[
            nonce.to_le_bytes().as_ref(),
            position_config.as_ref(),
            signer.as_ref(),
            ORDER_POSITION_SEED.as_bytes(),
        ],
        &program_id,
    )
    .0;

    let b = Pubkey::find_program_address(
        &[
            nonce.to_le_bytes().as_ref(),
            signer.as_ref(),
            ORDER_POSITION_SEED.as_bytes(),
        ],
        &program_id,
    )
    .0;

    println!(":::: {}", a);
    println!(":::: {}", b);

    Pubkey::find_program_address(
        &[
            nonce.to_le_bytes().as_ref(),
            position_config.as_ref(),
            signer.as_ref(),
            ORDER_POSITION_SEED.as_bytes(),
        ],
        &program_id,
    )
    .0
}
